File: playbook-java/src/main/java/io/infinitic/playbook/java/asyncMethod/workflows/AsyncMethodWorkflowImpl.java
package io.infinitic.playbook.java.asyncMethod.workflows;

import io.infinitic.playbook.java.Worker;
import io.infinitic.playbook.java.asyncMethod.services.RemoteService;
import io.infinitic.workflows.Deferred;
import io.infinitic.workflows.Workflow;

import java.time.LocalDateTime;
import static io.infinitic.workflows.DeferredKt.and;

public class AsyncMethodWorkflowImpl extends Workflow implements AsyncMethodWorkflow {

    private final AsyncMethodWorkflow self = getWorkflowById(AsyncMethodWorkflow.class, getWorkflowId());
    private final RemoteService remoteService = newService(RemoteService.class);

    @Override
    public void run() {
        Deferred<Void> deferred1 = dispatchVoid(self::remoteServiceRun, "1", 10L);
        Deferred<Void> deferred2 = dispatchVoid(self::remoteServiceRun, "2", 10L);
        Deferred<Void> deferred3 = dispatchVoid(self::remoteServiceRun, "3", 1L);

        and(deferred1, deferred2, deferred3).await();
        log("all completed");
    }

    @Override
    public void remoteServiceRun(String id, long input) {
        log(id + " starting");
        remoteService.run(id, input);
        log(id + " completed");
    }

    private void log(String msg) {
        inlineVoid(() -> System.out.println(
                LocalDateTime.now().format(Worker.formatter) +
                        " - Workflow " + getWorkflowId() + " (method " + getMethodId() + ") - " + msg
        ));
    }
}

File: playbook-kotlin/src/main/kotlin/io/infinitic/playbook/kotlin/asyncMethod/workflows/AsyncMethodWorkflowImpl.kt
package io.infinitic.playbook.kotlin.asyncMethod.workflows

import io.infinitic.playbook.kotlin.formatter
import io.infinitic.playbook.kotlin.asyncMethod.services.RemoteService
import io.infinitic.workflows.and
import io.infinitic.workflows.Workflow
import java.time.LocalDateTime

class AsyncMethodWorkflowImpl : Workflow(), AsyncMethodWorkflow {

    private val self = getWorkflowById(AsyncMethodWorkflow::class.java, workflowId);
    private val remoteService = newService(RemoteService::class.java)

    override fun run() {
        val deferred1 = dispatch(self::remoteServiceRun, "1", 10)
        val deferred2 = dispatch(self::remoteServiceRun, "2", 10)
        val deferred3 = dispatch(self::remoteServiceRun, "3", 1)

        (deferred1 and deferred2 and deferred3).await()
        log("all completed")
    }

    override fun remoteServiceRun(id: String, input: Long) {
        log("$id starting")
        remoteService.run(id, input)
        log("$id completed")
    }

    private fun log(msg: String) = inline {
        println("${LocalDateTime.now().format(formatter)} - Workflow $workflowId (method $methodId) - $msg")
    }
}